### ISSUE TITLE
OperationCanceledException when ingesting AuditMessages

### ISSUE THREAD

**Issue by neidi** [Particular/ServiceControl#4023]
#### Description
We recently upgraded ServiceControl from Version 4.33.1 via 4.33.2 to 5.0.5. Since then we have the problem that the bulk insert from the Audit instance fails multiple times per day and we need to follow the guide to manually resolve the problem.

Some key points to our system:
- 135M messages in the Audit database requiring 1.7TB of storage.
- retention period 1y
- running on Azure VM with disk that adheres to the requirements of 7500 IOPS

During our analysis with the extended log levels for Raven and ServiceControl we found out that:
- most of the time, the bulk insert for up to 30 messages takes between 10-200ms.
- we have multiple ingestions where the bulk insert for two messages took up to 50s.
- during high ingestion times (see previous point), RavenDB peaked at the memory level of > 14GB.
- before and after the exception occurs, we found activities of the required disk space check. We don't know if this could impact it as well, we just wanted to mention it.

#### Expected behavior
Ingestion succeeds without exception and ingestion times are low (<50ms on average per message).

#### Actual behavior
Ingestion fails with the following stack trace:

```
2024-03-21 11:42:23.7189|55|Warn|ServiceControl.Audit.Auditing.AuditPersister|Bulk insertion dispose failed|Raven.Client.Exceptions.Documents.BulkInsert.BulkInsertAbortedException: Failed to execute bulk insert ---> System.AggregateException: One or more errors occurred. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__118`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Http\RequestExecutor.cs:line 932
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ExecuteAsync>d__69`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 700
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<GetExceptionFromOperation>d__60.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 644
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 367
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 382
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<>c__DisplayClass36_0.<<-ctor>b__0>d.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 245
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Sparrow.Threading.DisposeOnceAsync`1.<DisposeAsync>d__4.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Sparrow\Threading\DisposeOnce.cs:line 222
```

#### Versions
ServiceControl 5.0.5
 

### Steps to reproduce

See description of the bug.

### Relevant log output

```shell
2024-03-21 11:42:15.3439|39|Debug|ServiceControl.Audit.Persistence.RavenDB.CustomChecks.CheckMinimumStorageRequiredForAuditIngestion|Check ServiceControl data drive space starting. Threshold 5%
2024-03-21 11:42:15.3439|39|Debug|ServiceControl.Audit.Persistence.RavenDB.CustomChecks.CheckMinimumStorageRequiredForAuditIngestion|Free space: 325043437568 | Total: 2199020105728 | Percent remaining 15%
2024-03-21 11:42:20.3752|39|Debug|ServiceControl.Audit.Persistence.RavenDB.CustomChecks.CheckMinimumStorageRequiredForAuditIngestion|Check ServiceControl data drive space starting. Threshold 5%
2024-03-21 11:42:20.3752|39|Debug|ServiceControl.Audit.Persistence.RavenDB.CustomChecks.CheckMinimumStorageRequiredForAuditIngestion|Free space: 325043372032 | Total: 2199020105728 | Percent remaining 15%
2024-03-21 11:42:23.7189|55|Warn|ServiceControl.Audit.Auditing.AuditPersister|Bulk insertion dispose failed|Raven.Client.Exceptions.Documents.BulkInsert.BulkInsertAbortedException: Failed to execute bulk insert ---> System.AggregateException: One or more errors occurred. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__118`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Http\RequestExecutor.cs:line 932
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ExecuteAsync>d__69`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 700
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<GetExceptionFromOperation>d__60.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 644
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 367
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 382
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<>c__DisplayClass36_0.<<-ctor>b__0>d.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 245
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Sparrow.Threading.DisposeOnceAsync`1.<DisposeAsync>d__4.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Sparrow\Threading\DisposeOnce.cs:line 222
2024-03-21 11:42:23.7189|55|Info|ServiceControl.Audit.Auditing.AuditIngestion|Ingesting messages failed|Raven.Client.Exceptions.Documents.BulkInsert.BulkInsertAbortedException: Failed to execute bulk insert ---> System.AggregateException: One or more errors occurred. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__118`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Http\RequestExecutor.cs:line 932
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ExecuteAsync>d__69`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 700
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<GetExceptionFromOperation>d__60.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 644
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 367
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<ThrowBulkInsertAborted>d__42.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 382
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.BulkInsert.BulkInsertOperation.<>c__DisplayClass36_0.<<-ctor>b__0>d.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Raven.Client\Documents\BulkInsert\BulkInsertOperation.cs:line 245
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Sparrow.Threading.DisposeOnceAsync`1.<DisposeAsync>d__4.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54092\src\Sparrow\Threading\DisposeOnce.cs:line 222
2024-03-21 11:42:23.7189|55|Debug|ServiceControl.Audit.Auditing.AuditIngestor|Ingesting 31 message contexts
2024-03-21 11:42:23.7189|55|Debug|ServiceControl.Audit.Auditing.AuditPersister|Batch size 31
2024-03-21 11:42:23.7189|55|Debug|ServiceControl.Audit.Auditing.AuditPersister|Emitting 0 commands and 0 control messages.
2024-03-21 11:42:23.7189|55|Debug|ServiceControl.Audit.Auditing.AuditPersister|0 commands and 0 control messages emitted.
```

#### Workarounds
Multiple times of manual reingestion of failed imports according to the documentation.

#### Additional information
Full log file available if needed.

**Comment by WilliamBZA** [comment]
Because the `OperationCanceledException` was [introduced in 5.0.0](https://github.com/Particular/ServiceControl/pull/3837), I wonder if the 50 seconds for 2 messages that you are periodically seeing is related to IOPs throttling... That would certainly go some way to explain why most ingestion takes milliseconds with the occasional ingestion taking multiple orders of magnitude more.

Could you reach out to Microsoft support and ask them if the virtual machine or disks were throttled around the time period where the ingestion was taking seconds instead of milliseconds? You should be able to get the timestamps from your log files to provide the correct timespan to Microsoft support.

If you're on premium you can also use disk metrics to check the [utilization metrics of IOPs](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-metrics#storage-io-utilization-metrics)

Otherwise, [this tool](https://github.com/Azure/throttled) also looks promising...

Please use one of these options to confirm if there is IOPs throttling happening? If we know that throttling is the trigger, then we can change how we do the bulk ingestion.

**Comment by neidi** [comment]
Hi @WilliamBZA 
Thanks for getting back to this.
Since we are on premium i used the disk metrics to check for the disk metrics you mentioned but there weren't any abnormalities or spikes during the times when the errors happened.
I will reach out to Microsoft when it happens again to check back with their support to maybe get further information for the issue.

**Comment by mikeminutillo** [comment]
@neidi We are adding a setting to allow you to configure the timeout in 5.0.6. We will let you know when it is available to download.

**Comment by WilliamBZA** [comment]
This will be released in 5.0.6, closing
